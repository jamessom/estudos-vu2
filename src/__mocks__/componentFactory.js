import { shallowMount, createLocalVue } from '@vue/test-utils';
import VueRouter from '@/router/index';

export default function componentFactory(options = {}) {
  return {
    wrapper: () => {
      const { component, componentOptions, withRouter } = options;
      const localvue = createLocalVue();

      if (withRouter) {
        localvue.use(VueRouter);

        return shallowMount(component, {
          localvue,
          ...componentOptions,
        });
      }

      return shallowMount(component, {
        localvue,
        ...componentOptions,
      });
    },
  };
}
