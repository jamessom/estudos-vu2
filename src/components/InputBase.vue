<template>
  <div class="field">
    <label v-if="this.label" class="label">{{ this.label }}</label>

    <div :class="this.controlClass">
      <input
        :class="this.inputClass"
        :type="this.inputType"
        :placeholder="this.inputPlaceholder"
        :value="computedValue"
        @input="onInput"
      />
    </div>
  </div>
</template>

<script>
export default {
  name: 'InputBase',
  props: {
    value: [Number, String],
    label: {
      type: String,
      default: '',
    },
    inputType: {
      type: String,
      default: 'text',
    },
    inputPlaceholder: String,
    controlClass: {
      type: String,
      default: 'control',
    },
    inputClass: {
      type: String,
      default: 'input',
    },
  },
  data() {
    return {
      newValue: this.value,
    };
  },
  computed: {
    computedValue: {
      get() {
        return this.newValue;
      },
      set(value) {
        this.newValue = value;
        this.$emit('input', value);
      },
    },
    valueLength() {
      if (typeof this.computedValue === 'string') {
        return this.computedValue.length;
      }

      if (typeof this.computedValue === 'number') {
        return this.computedValue.toString().length;
      }

      return 0;
    },
  },
  watch: {
    value(value) {
      this.newValue = value;
    },
  },
  methods: {
    onInput(event) {
      this.$nextTick(() => {
        if (event.target) {
          this.computedValue = event.target.value;
        }
      });
    },
  },
};
</script>
