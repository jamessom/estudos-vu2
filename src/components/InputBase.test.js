import componentFactory from '@/__mocks__/componentFactory';
import faker from 'faker';
import InputBase from './InputBase.vue';

describe('src/components/InputBase.vue', () => {
  describe('when component is mounted', () => {
    const snapshotTest = componentFactory({ component: InputBase });
    const wrapper = snapshotTest.wrapper();

    it('should be a vue instance', () => {
      expect(wrapper.isVueInstance()).toBeTruthy();
    });

    it('should be a InputBase component', () => {
      expect(wrapper.name()).toBe('InputBase');
    });

    it('should renders input element by default', () => {
      expect(wrapper.contains('input')).toBeTruthy();
    });

    it('should have a control class', () => {
      expect(wrapper.classes()).toContain('field');
    });

    it('should renders html correctly', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('create componente with default props', () => {
    const propsData = {
      inputPlaceholder: faker.random.words(),
    };

    const factory = componentFactory({
      component: InputBase,
      componentOptions: {
        propsData,
      },
    });

    const wrapperProps = factory.wrapper().props();

    it('should create a input component with default label', () => {
      expect(wrapperProps.label).toBe('');
    });

    it('should create a input component with default type', () => {
      expect(wrapperProps.inputType).toBe('text');
    });

    it('should create a input component with placeholder text', () => {
      expect(wrapperProps.inputPlaceholder).toBe(propsData.inputPlaceholder);
    });

    it('should create a input component with default control class', () => {
      expect(wrapperProps.controlClass).toBe('control');
    });

    it('should create a input component with default input class', () => {
      expect(wrapperProps.inputClass).toBe('input');
    });
  });

  describe('create components and change default props', () => {
    const propsData = {
      label: faker.random.word(),
      inputType: faker.helpers.randomize([
        'email', 'text', 'password', 'number',
      ]),
      inputPlaceholder: faker.random.words(),
      controlClass: faker.random.word(),
      inputClass: faker.random.word(),
    };

    const factory = componentFactory({
      component: InputBase,
      componentOptions: {
        propsData,
      },
    });

    const wrapperProps = factory.wrapper().props();

    it('should create a input component with label text', () => {
      expect(wrapperProps.label).toBe(propsData.label);
    });

    it('should create a input component with type', () => {
      expect(wrapperProps.inputType).toBe(propsData.inputType);
    });

    it('should create a input component with placeholder text', () => {
      expect(wrapperProps.inputPlaceholder).toBe(propsData.inputPlaceholder);
    });

    it('should create a input component with control class', () => {
      expect(wrapperProps.controlClass).toBe(propsData.controlClass);
    });

    it('should create a input component with input class', () => {
      expect(wrapperProps.inputClass).toBe(propsData.inputClass);
    });
  });
});
