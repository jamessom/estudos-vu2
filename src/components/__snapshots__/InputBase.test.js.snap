// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/InputBase.vue when component is mounted should renders html correctly 1`] = `
<div class="field">
  <!---->
  <div class="control"><input type="text" class="input"></div>
</div>
`;
